project/api/models.py

```diff
@@ -19,13 +19,65 @@
 
 from requests import get
 from django.core.exceptions import ValidationError
+from django.contrib.auth.models import User
 from django.contrib.gis.db import models
 from django.contrib.postgres.fields import ArrayField
+from django.db.models.signals import post_save
+from django.dispatch import receiver
 from ordered_model.models import OrderedModel
 from colorfield.fields import ColorField
 from simple_history.models import HistoricalRecords
 
 
+class Vote(models.Model):
+    """
+    Abstract class to store User's votes
+    """
+
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    vote = models.BooleanField()
+
+    class Meta:
+        abstract = True
+
+
+class NarrativeVote(Vote):
+    """
+    Stores votes for Narratives. Extends Vote model
+    """
+
+    narrative = models.ForeignKey("Narrative", on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ("narrative", "user")
```

Comment:
What will happen if some user decided to change his vote? Will the value be updated?

Reply:
In the ORM we can change  `NarrativeVote.objects.get(user={user}, narrative={narrative}).vote`. I guess we can set up filtering in the API for these parameters

Reply:
I did not get understand, can you rephrase it? What exactly should we filter via API requests?

I was asking about upserting new vote result.

example:
- _
Narrative B - rating 0;
- User-A upvoted Narrative B (submitting `true`)
Narrative B - rating 1;
- User-A downvoted Narrative B (submitting `false`)
Narrative B - rating -1;
- User-A took back his vote (submitting `undefined`)
Narrative B - rating 0;

Reply:
I envisioning this flow for the procedure you outlined:


* User-A upvoted Narrative B (submitting true)
    * `POST` to  `/api/narrative-votes/`
* User-A downvoted Narrative B (submitting false)
    * `PUT` to `/api/narrative-votes/[id of vote in first step]/`
* User-A took back his vote (submitting undefined)
    * `DELETE` to `/api/narrative-votes/[id of vote in first step]/`

And then, to retrieve the ID of the vote, I was thinking we should set up some filtering in the API by `?narrative` and `?user`. How do you feel about this flow?

Reply:
Well, that will require frontend to know about every vote and keep track of them.

I was thinking of less RESTful approach - using `narrative_id` as the key, not the `vote_id`.

* User-A upvoted Narrative B (submitting true)
  * POST to /api/narrative-votes/[narrative_id]/
* User-A downvoted Narrative B (submitting false)
  * POST to /api/narrative-votes/[narrative_id]/
* User-A took back his vote (submitting undefined)
  * POST to /api/narrative-votes/[narrative_id]/

But it is probably because I don't want to think about votes on the frontend

Reply:
Perhaps, as to not break RESTful guidelines too much, we could POST to /api/narrative-votes/ with `{ narrative, user, vote }` and this will update an existing reference if it already exists, using the django helper method [update_or_create](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#update-or-create). I can also configure a delete to be performed if `vote: null`, I think DRF strips `undefined` values in the serializer.

> And it would be great if it will be possible to return new vote results after POSTing vote

To clarify, are you thinking after the request something like `{ upvotes, downvotes }` could be returned for that specific narrative? I'm fine with that, and with returning those counters in /api/narratives/

Reply:
> with { narrative, user, vote }

Ok, that will work

I can fetch `/api/narratives/` after voting